### ISSUE TITLE
Speed up mcts

### ISSUE THREAD

**Issue by fqjin** [fqjin/2048NN#14]
Current selfplay speed is still too long. The biggest own-time bottlenecks are:

CPU (7 seconds), for some reason, loading the model takes much longer here (+3 sec)
  * `conv2d`: 31%
  * `move_batch`: 14%
  * `merge_row_batch`: 13%
  * `generate_tile`: 12%

GPU (15 seconds):
  * `move_batch`: 32%
  * `zeros`: 18%
  * `generate_tile`: 15%
  * `merge_row_batch`: 15%

**Comment by fqjin** [comment]
Performing network predictions on the GPU reduces the time of `conv2d`, but overall takes longer (9 sec) due to data transfer to GPU.

**Comment by fqjin** [comment]
I rewrote `generate_tile_batch` in 5271fabdcb4569c5c2af4630a04619bb31738c32, which is 5 times faster at a batch size of 200. Combined with some slight improvements to `mcts_nn`, I cut 1.5 seconds (20%) off CPU time and 4 seconds (25%) off GPU time:

CPU (5.7 sec):
  * `conv2d`:  33%
  * `move_batch`: 15%
  * `merge_row_batch`: 14%
  * `generate_tile_batch`: 7%

GPU (11.6 sec):
  * `move_batch`: 39%
  * `zeros`: 24% (tensor creation/transfer to GPU)
  * `merge_row_batch`: 16%
  * `generate_tile_batch`: 8%

**Comment by fqjin** [comment]
Minor optimizations in 508b59895c697519982947f30253ca667d8f0d63 (flip function dictionary, list comprehensions) significantly reduces own time of `move_batch`. Improves speed by 10% on CPU and 23% on GPU.

Note: Lines played are different between CPU and GPU due to different RNG (despite same seed).

CPU (5.1 sec):
* `conv2d`: 37%
* `move_batch`: 3%
* `merge-row-batch`: 15%
* `generate_tile_batch`: 7%

GPU (8.9 sec):
* `move_batch`: 7%
* `zeros`: 31%
* `merge-row-batch`: 21%
* `generate_tile_batch`: 11%

**Comment by fqjin** [comment]
The function `zeros` represents the initialization of the GPU, which is separate from actually running the mcts. I pre-initialize the GPU using `ones` to separate that time from the actual function times.

GPU (8.7 sec):
* `merge_row_batch`: 21%
* `generate_tile_batch`: 12%
* `Tensor.item`: 11%
* `Tensor.flip`: 7%
* `move_batch`: 7%

**Comment by fqjin** [comment]
I tried again to send the data to GPU for model prediction, and now it is 20% faster. Previous attempt I must have been counting the cuda initialization.

CPU --> GPU (4.0 sec):
* `merge_row_batch`: 20%
* `conv2d`: 15%
* `generate_tile_batch`: 9%

**Comment by fqjin** [comment]
Data transfer was not optimized in `mcts_nn.py`:
`preds = model.forward(torch.stack(boards).float().unsqueeze(1).cuda())` [code](https://github.com/fqjin/2048NN/blob/7467d3ec3ccc6b32c5ee4053c65814149a4f47ed/mcts_nn.py#L42)

`uint8` is cast to `float` before transferring to `cuda`. Casting is 2x faster on `cuda` and data transfer is 4x slower as `float`. The ideal order is `.cuda().float().unsqueeze(1)`. I recently changed the order in `game_dataset.py` (a5a0bb9fcecd42d719b43d81d9093e6b1c7beef2).

Before: 15.143 sec
After: 14.920 sec
Unfortunately not too much a speed up.

**Comment by fqjin** [comment]
Conservative strategy: returns the move count of the first line to die for each legal move. Search is then terminated immediately, removing the need for long playouts.
`mcts_nn`: 14885 ms
`mcts_nn_min`: 830 ms
18x faster. Argmax of the mcts results tends to agree for the most part, although the new method will be more sensitive to noise because it does not use averaging.

**Comment by fqjin** [comment]
With the new strategy, play is much faster. Selfplay games are generated in a few minutes, and multiple processes can be run simultaneously (limited by cores and GPU memory). Closing this issue as speed has already been improved quite a lot.
